BUG SparseDataFrame with dense Series (#19374)

datapythonista opened the pull request:
- [X] closes #19374
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

gfyoung commented on pandas/tests/sparse/frame/test_frame.py:
```diff
@@ -199,6 +199,31 @@ def test_constructor_from_series(self):
+    def test_constructor_from_unknown_type(self):
+        class Unknown:
+            pass
+        pytest.raises(TypeError, SparseDataFrame, Unknown())
```
Let's check error message too.

gfyoung commented on pandas/tests/sparse/frame/test_frame.py:
```diff
@@ -199,6 +199,31 @@ def test_constructor_from_series(self):
         # without sparse value raises error
         # df2 = SparseDataFrame([x2_sparse, y])
 
+    def test_constructor_from_dense_series(self):
```
Reference issue number under all of your added tests.

gfyoung commented:
@datapythonista : Looks pretty good so far.  Don't forget to add a `whatsnew` entry.

jreback commented on pandas/tests/sparse/frame/test_frame.py:
```diff
@@ -199,6 +199,31 @@ def test_constructor_from_series(self):
+        x = Series(np.random.randn(10000), name='a')
+        assert isinstance(x, Series)
+        df = SparseDataFrame(x)
+        assert isinstance(df, SparseDataFrame)
```
construct an expected SDF and compare

jreback replied:
construct a DataFrame and use .to_sparse() to construct an expected frame. we do not want to do all of these little checks, we already have well established comparison functions, e.g. ``tm.assert_sparse_equal`` for this

jreback commented on pandas/tests/sparse/frame/test_frame.py:
```diff
@@ -199,6 +199,31 @@ def test_constructor_from_series(self):
+        assert isinstance(df, SparseDataFrame)
+        assert df.columns == ['b']
+
+        # No column name available
```
same for the rest

jreback commented on pandas/core/sparse/frame.py:
```diff
@@ -95,6 +95,13 @@ def __init__(self, data=None, index=None, columns=None, default_kind=None,
         elif isinstance(data, DataFrame):
             mgr = self._init_dict(data, data.index, data.columns, dtype=dtype)
+        elif isinstance(data, Series):
+            if columns is None and data.name is None:
```
need a test to hit this case

jreback reviewed (changes requested):
pls add a whatsnew note as well.

datapythonista commented:
Thanks a lot for the comments. Sorry about the whatsnew, I added it when opening the PR, but forgot to add it to the commit.

@jreback, I didn't find a way to construct the expected `SparseDataFrame` without using the constructor that is being tested itself. But I added the comparison of the length and sum of the original series and the new `SparseDataFrame`, which I think should be enough (and it helped find a bug in my previous code).

Addressed all the other comments, let me know if you see anything else. Thanks!

jreback commented on pandas/core/sparse/frame.py:
```diff
@@ -95,6 +95,20 @@ def __init__(self, data=None, index=None, columns=None, default_kind=None,
+            elif len(columns) != 1:
+                raise ValueError('columns must be of length one '
+                                 'if data is of type Series')
+            mgr = self._init_dict(data.to_frame(columns[0]),
```
don't construct like this, actually make dict

jreback commented on pandas/core/sparse/frame.py:
```diff
@@ -95,6 +95,20 @@ def __init__(self, data=None, index=None, columns=None, default_kind=None,
+            if columns is None:
+                if data.name is None:
+                    raise ValueError('cannot pass a series '
+                                     'w/o a name or columns')
```
I don't think you need any of this, you can simply all ``to_manager`` with the column (in a list)

datapythonista commented:
Thanks for the feedback @jreback, I misunderstood what the columns argument had to do, and I was overcomplicating the code.

Now it simply creates the `DataFrame` with the `Series` name if it has one, or it uses 0 as the column if it doesn't. This seems to be consistent with the `Series` constructor and with creating a `SparseDataFrame` from a `DataFrame`, which ignores unknown columns.

I'm not 100% sure I understood what you mean by "actually contruct dic", please let me know if this new version doesn't address this too.

jreback commented on pandas/tests/sparse/frame/test_frame.py:
```diff
@@ -199,6 +199,33 @@ def test_constructor_from_series(self):
+        # GH 19393
+        # series with name
+        x = Series(np.random.randn(10000), name='a')
+        assert isinstance(x, Series)
```
you don't need this assert

jreback commented on pandas/tests/sparse/frame/test_frame.py:
```diff
@@ -199,6 +199,33 @@ def test_constructor_from_series(self):
+        # series with name
+        x = Series(np.random.randn(10000), name='a')
+        assert isinstance(x, Series)
+        res = SparseDataFrame(x)
```
I prefer result and expected

jreback commented on pandas/tests/sparse/frame/test_frame.py:
```diff
@@ -199,6 +199,33 @@ def test_constructor_from_series(self):
+        x = Series(np.random.randn(10000), name='a')
+        assert isinstance(x, Series)
+        res = SparseDataFrame(x)
+        assert isinstance(res, SparseDataFrame)
```
you don't need the assert

jreback commented on pandas/tests/sparse/frame/test_frame.py:
```diff
@@ -199,6 +199,33 @@ def test_constructor_from_series(self):
+        # series with no name
+        x = Series(np.random.randn(10000))
+        assert isinstance(x, Series)
+        res = SparseDataFrame(x)
```
same w.r.t naming & asserts

jreback reviewed (changes requested):
small edits, other lgtm. ping on green.

Commit:
BUG adding support for dense Series in the SparseDataFrame constructor, and providing useful error messages for other types (#19374)

datapythonista commented:
Thanks once again for the comments @jreback, seems that I followed the conventions of an old test. Addressed your comments, should be all right now.

jreback commented:
thanks @datapythonista
